lib/nrf9160_time/nrf9160_time.c

```diff
@@ -0,0 +1,284 @@
+	if (err) {
+		goto freeaddr;
+	}
+
+	err = sntp_query(&sntp_ctx, timeout, time);
+	if (err) {
+		sntp_close(&sntp_ctx);
+		goto freeaddr;
+	}
+
+freeaddr:
+	freeaddrinfo(addr);
+	return err;
+}
+
+static int get_time_NTP_server(void)
+{
+	int err;
+	int i = 0;
+
+	while (i < ARRAY_SIZE(servers)) {
+		err =  sntp_request_time(servers[i].server,
+					 K_SECONDS(5), &sntp_time);
+		if (err) {
+			LOG_DBG("Not getting time from NTP server %s, error %d",
+				log_strdup(servers[i].server), err);
+			LOG_DBG("Trying another address...");
+			i++;
+			continue;
+		}
+
+		LOG_DBG("Got time response from NTP server %s",
+			log_strdup(servers[i].server));
+		time_aux.date_time_utc = (s64_t)sntp_time.seconds * 1000;
+		time_aux.last_date_time_update = k_uptime_get();
+		return 0;
+	}
+
+	LOG_ERR("Not getting time from any NTP server");
+
+	return -ENODATA;
+}
+
+static int check_current_time(void)
+{
+	if (time_aux.last_date_time_update == 0 ||
+	    time_aux.date_time_utc == 0) {
+		LOG_DBG("Date time never set");
+		return -ENODATA;
+	}
+
+	if ((k_uptime_get() - time_aux.last_date_time_update) >
+	    CONFIG_NRF9160_TIME_UPDATE_INTERVAL * 1000) {
+		LOG_DBG("Current date time too old");
+		return -ENODATA;
+	}
+
+	return 0;
+}
+
+static void update_new_date_time(void)
+{
+	int err;
+
+	while (true) {
+		k_sem_take(&fetch_time_sem, K_FOREVER);
+
+		LOG_DBG("Updating date time UTC...");
+
+		err = check_current_time();
+		if (err == 0) {
+			LOG_DBG("Time successfully obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Current time not valid");
+
+#if defined(CONFIG_NRF9160_TIME_MODEM)
+		LOG_DBG("Fallback on cellular network time");
+
+		err = get_time_cellular_network();
+		if (err == 0) {
+			LOG_DBG("Time from cellular network obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Not getting cellular network time");
+#endif
+
+		LOG_DBG("Fallback on NTP server");
+
+		err = get_time_NTP_server();
+		if (err == 0) {
+			LOG_DBG("Time from NTP server obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Not getting time from NTP server");
+	}
+}
+
+K_THREAD_DEFINE(time_thread, CONFIG_NRF9160_TIME_THREAD_SIZE,
+		update_new_date_time, NULL, NULL, NULL,
+		K_HIGHEST_APPLICATION_THREAD_PRIO, 0, K_NO_WAIT);
+
+static void nrf9160_time_handler(struct k_work *work)
+{
+	k_sem_give(&fetch_time_sem);
+
+	LOG_DBG("New date time update in: %d seconds",
+		CONFIG_NRF9160_TIME_UPDATE_INTERVAL);
+
+	k_delayed_work_submit(&time_work,
+			      K_SECONDS(CONFIG_NRF9160_TIME_UPDATE_INTERVAL));
+}
+
+void nrf9160_time_init(void)
+{
+	k_delayed_work_init(&time_work, nrf9160_time_handler);
+	k_delayed_work_submit(&time_work, K_NO_WAIT);
+}
+
+void date_time_set(const struct tm *new_date_time)
+{
+	time_aux.last_date_time_update = k_uptime_get();
+	time_aux.date_time_utc = (s64_t)timeutil_timegm64(new_date_time) * 1000;
+}
+
+int uptime_to_unix_time_ms(const s64_t *uptime, s64_t *unix_time_ms)
```

Comment:
Since the uptime is a quite large integer I wanted to avoid the memory overhead associated with passing it in by value. How about removing the ``uptime`` all togheter and expecting the unix_time_ms to carry the ``uptime date``, essentially making this function to operate on a single pointer. [in, out]?